### PR TITLE
docker: Add support for volume binding and environment variable claim

### PR DESCRIPTION
**Pull request by vanloswang**
CC https://github.com/cockpit-project/cockpit/issues/4506

### CODE REVIEW

**Comment by stefwalter** [comment]
@vanloswang Nice work. Looks like one thing that's missing is the option to make a mount read-only or not. But I don't mind if we leave that for a later pull request.

File: pkg/docker/index.html [hunk 1]
```diff
@@ -496,6 +537,28 @@ <h4 class="modal-title">Run Image</h4>
...
+                <label>
+                  <input type="checkbox" name="mount-volumes" id="mount-volumes" checked="false"/>
+                  Mount container volumes
+                </label>
+                <div id="select-mounted-volumes" class="containers-run-volumemount containers-run-inline">
+                </div>
+              </td>
+            </tr>
+            <tr>
+              <td><label class="control-label" for="claim-envvars" translatable="yes">Env Vars</label></td>
```
Comment (stefwalter) [inline pkg/docker/index.html]: This line would be clearer if it said 'Environment'

File: pkg/docker/index.html [hunk 2]
```diff
@@ -496,6 +537,28 @@ <h4 class="modal-title">Run Image</h4>
...
+              <td colspan="4">
+                <label>
+                  <input type="checkbox" name="claim-envvars" id="claim-envvars" checked="false"/>
+                  Claim container environment variables
```
Comment (stefwalter) [inline pkg/docker/index.html]: This line would be clearer if it said 'Set environment variables'. In addition the text here should be inside a ```<span translatable="yes">```

**Comment by stefwalter** [comment]
When I start a container with a volume mount ... and then go to start a new container ... the dialog shows me the last choices, however the relevant checkbox is not selected. I think the dialog should start with what the image config describes, rather that the previous choices.

screenshot from 2016-07-12 23-05-00

**Comment by vanloswang** [comment]
@stefwalter yes, it should read configs from the image. I will try it and commit it here or in later pull request. Thx.

**Comment by vanloswang** [comment]
@stefwalter Do you know how to get 'Binds' info from 'PageRunImage.image_info'? Now I can get 'Env' from 'PageRunImage.image_info.Config.Env', but I can not get 'Binds' from 'PageRunImage.image_info.Config.HostConfig.Binds'.

**Comment by vanloswang** [comment]
@stefwalter Got it form 'PageRunImage.image_info.Config.Volumes'.

**Comment by vanloswang** [comment]
@stefwalter all feature done, need to check again. Thx.

**Comment by andreasn** [comment]
I haven't tried this in action (yet), but looking at the screenshot, the "to host path" and the second input ends up on a new line. Is it possible to keep it on the same line as the first input?
Looks really good apart from that. Thanks for hacking on this!

**Comment by vanloswang** [comment]
@andreasn Thx for your advice, the latest commit contain css fix.
@stefwalter need check again.

**Comment by andreasn** [comment]
This dropdown is a bit too wide. I wonder what's triggering that.

screenshot from 2016-07-18 13-19-47

File: pkg/docker/run.js [hunk 3]
```diff
@@ -184,6 +214,43 @@ define([
                 $('#expose-ports').prop('checked', false);
             }
 
+            /* delete any old volume binding entries */
+            var volumebinding = $('#select-mounted-volumes');
```
Comment (dperpeet) [inline pkg/docker/run.js]: variable names should be either camelCase or old_style

File: pkg/docker/run.js [hunk 4]
```diff
@@ -184,6 +214,43 @@ define([
...
+            if (volumebinding.children().length > 0) {
+                $('#mount-volumes').prop('checked', true);
+                /* make sure the volumes are visible */
+                volumebinding.show();
+            } else {
+                $('#mount-volumes').prop('checked', false);
+            }
+
+            /* delete any old env var claiming entries */
+            var envvarclaiming = $('#select-claimed-envvars');
```
Comment (dperpeet) [inline pkg/docker/run.js]: variable names should be either camelCase or old_style

**Comment by dperpeet** [comment]
this will be in cockpit-0.116

**Pull request closed without merging**